examples/seq2seq/seq2seq_trainer.py

```diff
@@ -217,7 +217,12 @@ def prediction_step(
         }
 
         if self.args.predict_with_generate and not self.args.prediction_loss_only:
-            generated_tokens = self.model.generate(
+            if self.args.deepspeed:
+                model = self.model.module
+            else:
+                model = self.model
```

Comment:
I don't like that `deepspeed` breaks the current assumption that `self.model` is always a reference to the proper model. We should fix that.

Reply:
So you're suggesting that we do the following:
```
if self.args.deepspeed:
    self.model = model.module
else:
    self.model = model
```
and then stash the deepspeed wrapped module somewhere else in `model` and use that during `backward`?

Perhaps we need to allocate a slot for the wrapped models so that the user will always be able to reach for one or the other?

So in case of DDP `self.wrapped_model` will be DDP wrapped model - in case of deepspeed it'd be its wrapper

Reply:
We can add `self.wrapped_model` for the wrapped model, I have nothing against that! I'd store the fully wrapped model in the case of deepspeed since that is the model we expect to use in the forward pass.

Reply:
well, actually, it's more complicated. `self.wrapped_model` will be `DDP(Deepspeed(Transformers Model))` - so depending on the need sometimes we will work with  `self.wrapped_model` (DDP) and other times  `self.wrapped_model.module` (DS).

Done.